runtime/vm/ExtendedMessageNPE.cpp

```diff
@@ -0,0 +1,201 @@
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "cfreader.h"
+#include "stackwalk.h"
+#include "ut_j9vm.h"
+
+extern "C" {
+
+char*
+getCompleteNPEMessage(J9VMThread *vmThread, U_8 *bcCurrentPtr, J9ROMClass *romClass, const char *npeCauseMsg)
+{
+
+	char *npeMsg = NULL;
+	UDATA msgLen = 0;
+	const char *nlsMessage = NULL;
+	U_8 bcCurrent = *bcCurrentPtr;
+
+	PORT_ACCESS_FROM_VMC(vmThread);
+	Trc_VM_GetCompleteNPEMessage_Entry(vmThread, bcCurrentPtr, bcCurrent, romClass, npeCauseMsg);
+	if (((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload))
+		|| ((bcCurrent >= JBiastore) && (bcCurrent <= JBsastore))
+	) {
+		const char *elementType = NULL;
+		switch (bcCurrent) {
+		case JBiaload: /* FALLTHROUGH */
+		case JBiastore:
+			elementType = "int";
+			break;
+		case JBlaload: /* FALLTHROUGH */
+		case JBlastore:
+			elementType = "long";
+			break;
+		case JBfaload: /* FALLTHROUGH */
+		case JBfastore:
+			elementType = "float";
+			break;
+		case JBdaload: /* FALLTHROUGH */
+		case JBdastore:
+			elementType = "double";
+			break;
+		case JBaaload: /* FALLTHROUGH */
+		case JBaastore:
+			elementType = "object";
+			break;
+		case JBbaload: /* FALLTHROUGH */
+		case JBbastore:
+			elementType = "byte/boolean";
+			break;
+		case JBcaload: /* FALLTHROUGH */
+		case JBcastore:
+			elementType = "char";
+			break;
+		case JBsaload: /* FALLTHROUGH */
+		case JBsastore:
+			elementType = "short";
+			break;
+		default:
+			Trc_VM_GetCompleteNPEMessage_Unreachable(vmThread, bcCurrent);
+		}
+		if ((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload)) {
+#define VM_NPE_ALOAD "Cannot load from %s array"
+			nlsMessage = VM_NPE_ALOAD;
+#undef VM_NPE_ALOAD
+		} else {
+#define VM_NPE_ASTORE "Cannot store to %s array"
+			nlsMessage = VM_NPE_ASTORE;
+#undef VM_NPE_ASTORE
+		}
+		msgLen = j9str_printf(PORTLIB, NULL, 0, nlsMessage, elementType);
+		/* msg NULL check omitted since str_printf accepts NULL (as above) */
+		npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
+		j9str_printf(PORTLIB, npeMsg, msgLen, nlsMessage, elementType);
+	} else {
+		switch (bcCurrent) {
+		case JBarraylength:
+#define VM_NPE_ARRAYLENGTH "Cannot read the array length"
+			msgLen = j9str_printf(PORTLIB, NULL, 0, VM_NPE_ARRAYLENGTH);
+			npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
+			j9str_printf(PORTLIB, npeMsg, msgLen, VM_NPE_ARRAYLENGTH);
+#undef VM_NPE_ARRAYLENGTH
```

Comment:
I'm unclear on why a define is preferable to local variable here.  The local seems to have less "noise" when reading the code
```suggestion
		case JBarraylength:
			const char *VM_NPE_ARRAYLENGTH = "Cannot read the array length";
			msgLen = j9str_printf(PORTLIB, NULL, 0, VM_NPE_ARRAYLENGTH);
			npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
			j9str_printf(PORTLIB, npeMsg, msgLen, VM_NPE_ARRAYLENGTH);
```

Reply:
This comes from the thought (and becomes a habit later) in early days that the defines save the memory defining local variables. It is negligible and probably covered by compiler optimization already.
Agreed, the local reads better.

Reply:
Make the local const and it shouldn't matter